File: backend/app/routes/coding.py
"""
Coding round endpoints (Day 3).

Thin HTTP layer over app.services.coding_service — mirrors
app.routes.aptitude's shape exactly: request/response validation via
Pydantic schemas, session lookup via the existing get_db dependency, and
CodingServiceError -> HTTPException translation. All actual state machine /
generation / execution / grading logic lives in the service and the
agents/ package.
"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.models.schemas import (
    CodeRunRequest,
    CodeRunResponse,
    CodeSubmitRequest,
    CodeSubmitResponse,
    CodingResultResponse,
    CodingStateResponse,
)
from app.services.coding_service import (
    CodingServiceError,
    get_current,
    get_result,
    run_code,
    start_coding,
    submit_code,
)

router = APIRouter(prefix="/api/sessions/{session_id}/coding", tags=["coding"])


@router.post("/start", response_model=CodingStateResponse)
async def start(session_id: str, db: Session = Depends(get_db)):
    try:
        return await start_coding(db, session_id)
    except CodingServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/current", response_model=CodingStateResponse)
def current(session_id: str, db: Session = Depends(get_db)):
    try:
        return get_current(db, session_id)
    except CodingServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.post("/run", response_model=CodeRunResponse)
async def run(session_id: str, payload: CodeRunRequest, db: Session = Depends(get_db)):
    try:
        return await run_code(db, session_id, payload.code)
    except CodingServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.post("/submit", response_model=CodeSubmitResponse)
async def submit(session_id: str, payload: CodeSubmitRequest, db: Session = Depends(get_db)):
    try:
        return await submit_code(db, session_id, payload.code)
    except CodingServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc


@router.get("/result", response_model=CodingResultResponse)
def result(session_id: str, db: Session = Depends(get_db)):
    try:
        return get_result(db, session_id)
    except CodingServiceError as exc:
        raise HTTPException(status_code=exc.status_code, detail=str(exc)) from exc

File: frontend/src/api/coding.js
const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000'

async function request(path, options = {}) {
  const res = await fetch(`${API_BASE_URL}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...options,
  })
  if (!res.ok) {
    let detail = `HTTP ${res.status}`
    try {
      const body = await res.json()
      detail = body.detail || detail
    } catch {
      // response body wasn't JSON — keep the generic status message
    }
    const error = new Error(detail)
    error.status = res.status
    throw error
  }
  return res.json()
}

export function startCoding(sessionId) {
  return request(`/api/sessions/${sessionId}/coding/start`, { method: 'POST' })
}

export function getCodingCurrent(sessionId) {
  return request(`/api/sessions/${sessionId}/coding/current`)
}

export function runCode(sessionId, code) {
  return request(`/api/sessions/${sessionId}/coding/run`, {
    method: 'POST',
    body: JSON.stringify({ code }),
  })
}

export function submitCode(sessionId, code) {
  return request(`/api/sessions/${sessionId}/coding/submit`, {
    method: 'POST',
    body: JSON.stringify({ code }),
  })
}

export function getCodingResult(sessionId) {
  return request(`/api/sessions/${sessionId}/coding/result`)
}
